simapp/abci.go

```diff
@@ -2,53 +2,161 @@ package simapp
+
+		// Validate VE sigs and check whether they have 2/3+ majority
+		hasTwoThirdsSigs := true
+		if err := app.StakingKeeper.ValidateVoteExtensions(ctx, ctx.BlockHeight(), ctx.ChainID(), req.LocalLastCommit.Votes, req.LocalLastCommit.Round); err != nil {
+			fmt.Println("PrepareProposal: Error occurred while validating VEs: ", err)
+			hasTwoThirdsSigs = false
 		}
 
+		var txs [][]byte
+		rawTxs := req.Txs
+		// Only include VE if there's majority
+		if hasTwoThirdsSigs {
+			fmt.Println("PrepareProposal: INJECTING VEs!")
+			// VoteExtensions will be encoded before other txs
+			extVoteInfo, err := json.Marshal(req.LocalLastCommit.Votes)
+			if err != nil {
+				return nil, err
+			}
+			txs = append(txs, extVoteInfo)
+		}
+
+		// encode the txs
+		txs = append(txs, rawTxs...)
+		return &abci.ResponsePrepareProposal{Txs: txs}, nil
+
+	}
+
+}
+
+// Verify whether the proposal has VEs since a malicious proposer could censor votes.
+// Check for 2/3+ VE sigs and reject the proposal in case we don't have a majority.
+// A downside is liveness can get negatively affected.
+func (app *SimApp) NewProcessProposalHandler() sdk.ProcessProposalHandler {
+	return func(ctx sdk.Context, req *abci.RequestProcessProposal) (*abci.ResponseProcessProposal, error) {
+		if req.Height > 5 {
+			var extVoteInfo []abci.ExtendedVoteInfo
+			// Extract ExtendedVoteInfo from txs (encoded at the beginning)
+			bz := req.Txs[0]
+			if err := json.Unmarshal(bz, &extVoteInfo); err != nil {
+				fmt.Println("Error occurred while decoding ExtendedVoteInfo: %w", err)
+			}
+			// Validate VE sigs and check whether they have 2/3+ majority
+			hasTwoThirdsSigs := true
+			// FIXME
+			if err := app.StakingKeeper.ValidateVoteExtensions(ctx, ctx.BlockHeight(), ctx.ChainID(), extVoteInfo, req.ProposedLastCommit.Round); err != nil {
+				hasTwoThirdsSigs = false
+			}
+
+			if !hasTwoThirdsSigs {
+				return &abci.ResponseProcessProposal{Status: abci.ResponseProcessProposal_REJECT}, nil
+			}
+		}
+
+		return &abci.ResponseProcessProposal{Status: abci.ResponseProcessProposal_ACCEPT}, nil
+	}
+
+}
+
+// ExtendVote -> Fetch data from buffer -> Vote on it accordingly -> Store result.
+func (h *VoteExtensionHandler) ExtendVote() sdk.ExtendVoteHandler {
+	return func(ctx sdk.Context, req *abci.RequestExtendVote) (*abci.ResponseExtendVote, error) {
+		fmt.Println("EXTENDING VOTE!", "height: ", ctx.BlockHeight())
 		ve := VoteExtension{
 			Hash:   req.Hash,
 			Height: req.Height,
-			Data:   buf,
+			Result: Vote_YES,
+		}
+
+		// txs are encoded after ExtendedVoteInfo
+		txs := req.Txs[1:]
+		for _, rawTx := range txs {
+			tx, err := h.app.TxDecode(rawTx)
+			if err != nil {
+				return nil, fmt.Errorf("Error occurred while decoding raw tx bytes: %w", err)
+			}
+			msgs := tx.GetMsgs()
+			for _, msg := range msgs {
+				switch msg.(type) {
+				// TODO: Redirect ExtendVote calls to individual module
+				case *stakingtypes.MsgEventRecord:
```

Comment:
Not for when implementing: handle or forbid having multiple MsgEventRecord/SideTx in the same block.